optional_plugins/varianter_cit/avocado_varianter_cit/Solver.py

```diff
@@ -107,18 +110,36 @@ def clean_hash_table(self, combination_matrix, t_value):
                     for key in itertools.product(*value_array):
                         combination_matrix.del_cell(c, key)
 
-    def clean_data_matrix(self, data_matrix, parameter=None):
-        if parameter is None:
-            for constraint in self.constraints:
-                if len(constraint) == 1:
-                    constraint = constraint[0]
-                    data_matrix[constraint.name].remove(constraint.value)
-        else:
-            parameter_constraint = self.parameters[parameter["name"]].constraints[parameter["value"]]
-            for constraint in parameter_constraint:
-                constraint = constraint[0]
-                try:
-                    data_matrix[constraint.name].remove(constraint.value)
-                except ValueError:
-                    # this value was already deleted
-                    pass
+    def get_possible_values(self, row, parameter):
+        """
+        Computes all possible values for the given parameter based on
+        constraints and already picked values of other parameters
+        :param row: row with picked values. -1 means an unpicked value.
+        :type row: list
+        :param parameter: index of the parameter
+         whose we want to know the values
+        :type parameter: int
+        :return: all possible values for the given parameter
+        :rtype: list
+        """
+
+        def is_permitted_value(one_value_constraints):
+            if len(one_value_constraints) == 0:
+                return False
+
+            for constraints in one_value_constraints:
+                is_ok = False
+                for constraint in constraints:
+                    if row[constraint.name] != constraint.value:
+                        is_ok = True
+                        break
+                if not is_ok:
+                    return False
+            return True
+
+        possible_values = []
+        parameter = self.parameters[parameter]
+        for value, constraints in enumerate(parameter.constraints):
+            if constraints is None or is_permitted_value(constraints):
+                possible_values.append(value)
+        return possible_values
```

Comment:
Just a suggestion, maybe we could replace this code by a `filter`

Reply:
That is a great idea, I will use that thanks.